src/components/page/LoopState.ts

```diff
@@ -0,0 +1,68 @@
+import FrameData from 'src/sketch/FrameData';
+
+export default class LoopState {
+  finished: boolean;
+
+  animationFrameRequest?: number;
+
+  frameData: FrameData;
+
+  paused: boolean;
+
+  startTime: number;
+  lastTime: number;
+
+  constructor() {
+    this.finished = false;
+
+    this.frameData = {
+      frame: 0,
+      frameTime: 0,
+      totalTime: 0,
+    };
+
+    this.paused = false;
+
+    const start = Date.now();
+    this.startTime = start;
+    this.lastTime = start;
+  }
+
+  togglePause() {
+    this.paused = !this.paused;
+  }
+
+  nextFrame() {
+    if (this.paused || this.finished) {
+      return false;
+    }
+
+    const requestTime = Date.now();
+    const ticks = requestTime - this.lastTime;
+    if (ticks > 10) {
+      this.frameData.frame += 1;
+      this.frameData.frameTime = ticks;
+      this.frameData.totalTime += ticks;
+
+      this.lastTime = requestTime;
+
+      return true;
+    }
+    return false;
+  }
+
+  restart() {
+    this.finished = false;

```

Comment:
this repeats the content from constructor. should look for a way to accomplish this without repeating the code.